lua/luasnip/loaders/from_snipmate.lua

```diff
@@ -107,60 +130,50 @@ local function filter(exclude, include)
 	end
 end
 
-local function get_paths(root)
-	local ret = {}
-	local files = Path.scandir(root, "file")
-	for _, file in ipairs(files) do
-		local name, ext = Path.basename(file, true)
-		if ext == "snippets" then
-			table.insert(ret, { path = Path.join(root, file), ft = name })
-		end
-	end
-	local dirs = Path.scandir(root, "directory")
-	for _, dir in ipairs(dirs) do
-		local name = dir
-		files = Path.scandir(Path.join(root, dir), "file")
-		for _, file in ipairs(files) do
-			if vim.endswith(file, ".snippets") then
-				table.insert(
-					ret,
-					{ path = Path.join(root, dir, file), ft = name }
-				)
-			end
-		end
+local M = {}
+
+local function normarize_paths(paths)
+	if not paths then
+		paths = vim.api.nvim_get_runtime_file("snippets", true)
+	elseif type(paths) == "string" then
+		paths = vim.split(paths, ",")
 	end
-	return ret
-end
 
-local function load_snippet_folder(root, opts)
-	local paths = get_paths(root)
-	local ft_filter = filter(opts.exclude, opts.include)
+	paths = vim.tbl_map(Path.expand, paths)
+	paths = util.deduplicate(paths)
 
-	for _, v in ipairs(paths) do
-		local path, ft = v.path, v.ft
-		if ft_filter(ft) then
-			load_snippet_file(ft, path)
+	cache.ft_paths = get_ft_paths(paths)
+end
+
+function M._load(ft)
+	local snippets = {}
+	for _, path in ipairs(cache.ft_paths[ft]) do
+		local snippet, extends = load_snippet_file(path)
+		vim.list_extend(snippets, snippet)
+		for _, extend in ipairs(extends) do
+			vim.list_extend(snippets, M._load(extend))
 		end
 	end
+	return snippets
 end
 
-local M = {}
-
 function M.load(opts)
 	opts = opts or {}
-	opts.exclude = opts.exclude or {}
 
-	if not opts.paths then
-		opts.paths = vim.api.nvim_get_runtime_file("snippets", true)
-	elseif type(opts.paths) == "string" then
-		opts.paths = vim.split(opts.paths, ",")
+	if not opts.is_lazy then
+		normarize_paths(opts.paths)
 	end
 
-	opts.paths = vim.tbl_map(Path.expand, opts.paths)
-	opts.paths = util.deduplicate(opts.paths)
+	local ft_filter = filter(opts.exclude, opts.include)
 
-	for _, path in ipairs(opts.paths) do
-		load_snippet_folder(path, opts)
+	for ft, _ in pairs(cache.ft_paths) do
+		if ft_filter(ft) then
+			local snippets = M._load(ft)
+			local lang_snips = ls.snippets[ft] or {}
+			ls.snippets[ft] = vim.list_extend(lang_snips, snippets)
+			session.latest_load_ft = ft
+			vim.cmd("do User LuasnipSnippetsAdded")
```

Comment:
Could we store which snippet-files were added to which filetype and then skip duplicates here?

Reply:
I don't think snippet-files necessarily corresponds to a single filetype.
I believe that the extends syntax is only expected to work for .snippets.
For example, if you have extends javascript in typescript.snippets, and you open .ts and then open .js, then javascript.snippets should be referenced twice. So I'm caching it.

Reply:
Ah, I was thinking completely wrong, nvm.